File: sdks/python/norikv/_internal/retry.py
"""Retry logic with exponential backoff for transient failures."""

import asyncio
import random
from typing import Callable, TypeVar, Optional, Set, Type

from norikv.errors import (
    NoriKVError,
    UnavailableError,
    DeadlineExceededError,
    NotLeaderError,
    RetryExhaustedError,
)
from norikv.types import RetryConfig

T = TypeVar("T")


class RetryPolicy:
    """Implements retry logic with exponential backoff and jitter."""

    # Errors that are retryable
    RETRYABLE_ERRORS: Set[Type[NoriKVError]] = {
        UnavailableError,
        DeadlineExceededError,
        NotLeaderError,
    }

    def __init__(self, config: RetryConfig):
        """Initialize the retry policy.

        Args:
            config: Retry configuration
        """
        self._config = config

    async def execute_with_retry(
        self,
        operation: Callable[[], T],
        operation_name: str = "operation",
    ) -> T:
        """Execute an operation with retry logic.

        Args:
            operation: Async function to execute
            operation_name: Name of the operation for error messages

        Returns:
            The result of the operation

        Raises:
            RetryExhaustedError: If all retry attempts are exhausted
            NoriKVError: If a non-retryable error occurs
        """
        attempt = 0
        last_error: Optional[Exception] = None

        while attempt < self._config.max_attempts:
            try:
                # Execute the operation
                result = await operation()
                return result

            except Exception as e:
                last_error = e
                attempt += 1

                # Check if error is retryable
                if not self._is_retryable(e):
                    raise

                # Check if we've exhausted retries
                if attempt >= self._config.max_attempts:
                    raise RetryExhaustedError(
                        f"{operation_name} failed after {attempt} attempts",
                        attempts=attempt,
                        last_error=str(e),
                    )

                # Calculate backoff delay
                delay = self._calculate_backoff(attempt)

                # Wait before retry
                await asyncio.sleep(delay)

        # Should not reach here, but handle it anyway
        raise RetryExhaustedError(
            f"{operation_name} failed after {attempt} attempts",
            attempts=attempt,
            last_error=str(last_error) if last_error else "Unknown error",
        )

    def _is_retryable(self, error: Exception) -> bool:
        """Check if an error is retryable.

        Args:
            error: The error to check

        Returns:
            True if the error should be retried
        """
        # Check if error is in retryable set
        for retryable_type in self.RETRYABLE_ERRORS:
            if isinstance(error, retryable_type):
                return True

        return False

    def _calculate_backoff(self, attempt: int) -> float:
        """Calculate backoff delay with exponential backoff and jitter.

        Args:
            attempt: The attempt number (1-indexed)

        Returns:
            Delay in seconds
        """
        # Base delay with exponential backoff
        base_delay = self._config.initial_delay_ms / 1000.0
        delay = base_delay * (self._config.backoff_multiplier ** (attempt - 1))

        # Cap at max delay
        max_delay = self._config.max_delay_ms / 1000.0
        delay = min(delay, max_delay)

        # Add jitter (random factor between 0.5 and 1.5)
        if self._config.jitter:
            jitter_factor = 0.5 + random.random()  # 0.5 to 1.5
            delay *= jitter_factor

        return delay


class CircuitBreaker:
    """Simple circuit breaker to prevent cascading failures."""

    def __init__(
        self,
        failure_threshold: int = 5,
        recovery_timeout: float = 60.0,
        half_open_max_requests: int = 3,
    ):
        """Initialize the circuit breaker.

        Args:
            failure_threshold: Number of failures before opening circuit
            recovery_timeout: Seconds to wait before trying to recover
            half_open_max_requests: Max requests to allow in half-open state
        """
        self._failure_threshold = failure_threshold
        self._recovery_timeout = recovery_timeout
        self._half_open_max_requests = half_open_max_requests

        self._state = "closed"  # closed, open, half-open
        self._failure_count = 0
        self._last_failure_time: Optional[float] = None
        self._half_open_requests = 0
        self._lock = asyncio.Lock()

    async def execute(
        self,
        operation: Callable[[], T],
        fallback: Optional[Callable[[], T]] = None,
    ) -> T:
        """Execute an operation through the circuit breaker.

        Args:
            operation: Async function to execute
            fallback: Optional fallback function if circuit is open

        Returns:
            The result of the operation or fallback

        Raises:
            UnavailableError: If circuit is open and no fallback provided
        """
        async with self._lock:
            # Check circuit state
            if self._state == "open":
                # Check if we should transition to half-open
                if self._should_attempt_recovery():
                    self._state = "half-open"
                    self._half_open_requests = 0
                else:
                    # Circuit is open
                    if fallback:
                        return await fallback()
                    raise UnavailableError(
                        "Circuit breaker is open",
                        retry_after_ms=int(self._recovery_timeout * 1000),
                    )

            # Allow request if closed or half-open (within limit)
            if self._state == "half-open":
                if self._half_open_requests >= self._half_open_max_requests:
                    if fallback:
                        return await fallback()
                    raise UnavailableError(
                        "Circuit breaker is half-open (max requests reached)",
                        retry_after_ms=1000,
                    )
                self._half_open_requests += 1

        # Execute operation outside the lock
        try:
            result = await operation()

            # Success - update state
            async with self._lock:
                if self._state == "half-open":
                    self._state = "closed"
                self._failure_count = 0
                self._last_failure_time = None

            return result

        except Exception as e:
            # Failure - update state
            async with self._lock:
                self._failure_count += 1
                self._last_failure_time = asyncio.get_event_loop().time()

                if self._failure_count >= self._failure_threshold:
                    self._state = "open"

            raise

    def _should_attempt_recovery(self) -> bool:
        """Check if enough time has passed to attempt recovery.

        Returns:
            True if we should try to recover
        """
        if self._last_failure_time is None:
            return True

        current_time = asyncio.get_event_loop().time()
        time_since_failure = current_time - self._last_failure_time

        return time_since_failure >= self._recovery_timeout

    def get_state(self) -> str:
        """Get the current circuit breaker state.

        Returns:
            Current state: "closed", "open", or "half-open"
        """
        return self._state

    def reset(self) -> None:
        """Reset the circuit breaker to closed state."""
        self._state = "closed"
        self._failure_count = 0
        self._last_failure_time = None
        self._half_open_requests = 0

File: sdks/typescript/src/retry.ts
/**
 * Retry logic with exponential backoff for NoriKV client.
 *
 * This module handles:
 * - Retrying failed requests with exponential backoff
 * - Distinguishing retryable vs non-retryable errors
 * - Handling NOT_LEADER redirects
 * - Respecting idempotency semantics
 */

import {
  NoriKVError,
  NotLeaderError,
  UnavailableError,
  DeadlineExceededError,
  ConnectionError,
  RetryExhaustedError,
  AlreadyExistsError,
  VersionMismatchError,
  InvalidArgumentError,
} from '@norikv/client/errors';
import type { RetryConfig } from '@norikv/client/types';

/**
 * Context for a retry operation.
 */
export interface RetryContext {
  /** Current attempt number (1-indexed) */
  attempt: number;

  /** Last error encountered */
  lastError?: Error;

  /** Whether this operation is idempotent */
  isIdempotent: boolean;

  /** Idempotency key (if provided) */
  idempotencyKey?: string;
}

/**
 * Default retry configuration.
 */
export const DEFAULT_RETRY_CONFIG: Required<RetryConfig> = {
  maxAttempts: 3,
  initialDelayMs: 10,
  maxDelayMs: 1000,
  backoffMultiplier: 2,
  jitterMs: 100,
  retryOnNotLeader: true,
};

/**
 * Determine if an error is retryable.
 */
export function isRetryable(error: Error, context: RetryContext): boolean {
  // Non-retryable errors
  if (error instanceof InvalidArgumentError) {
    return false; // Client error - won't succeed on retry
  }

  if (error instanceof AlreadyExistsError) {
    return false; // Conflict - retry won't help
  }

  if (error instanceof VersionMismatchError) {
    return false; // Optimistic locking failure - caller must handle
  }

  // Retryable errors
  if (error instanceof NotLeaderError) {
    return true; // Can retry on the correct leader
  }

  if (error instanceof UnavailableError) {
    return true; // Temporary unavailability
  }

  if (error instanceof DeadlineExceededError) {
    return context.isIdempotent; // Only retry timeouts if idempotent
  }

  if (error instanceof ConnectionError) {
    return true; // Network error - might succeed on retry
  }

  if (error instanceof NoriKVError) {
    // Generic NoriKV error - retry if idempotent
    return context.isIdempotent;
  }

  // Unknown error - only retry if idempotent
  return context.isIdempotent;
}

/**
 * Compute the delay before the next retry attempt.
 */
export function computeRetryDelay(
  attempt: number,
  config: Required<RetryConfig>
): number {
  // Exponential backoff: delay = initialDelay * (multiplier ^ attempt)
  const exponentialDelay = config.initialDelayMs * Math.pow(config.backoffMultiplier, attempt - 1);

  // Cap at max delay
  const cappedDelay = Math.min(exponentialDelay, config.maxDelayMs);

  // Add jitter: random value in [-jitterMs, +jitterMs]
  const jitter = (Math.random() - 0.5) * 2 * config.jitterMs;

  return Math.max(0, cappedDelay + jitter);
}

/**
 * Sleep for a given duration.
 */
export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Retry an operation with exponential backoff.
 *
 * @param operation - The operation to retry
 * @param config - Retry configuration
 * @param context - Retry context
 * @returns Result of the operation
 * @throws RetryExhaustedError if max attempts exceeded
 */
export async function withRetry<T>(
  operation: (ctx: RetryContext) => Promise<T>,
  config: RetryConfig,
  context: Omit<RetryContext, 'attempt'>
): Promise<T> {
  const fullConfig: Required<RetryConfig> = {
    ...DEFAULT_RETRY_CONFIG,
    ...config,
  };

  let lastError: Error | undefined;

  for (let attempt = 1; attempt <= fullConfig.maxAttempts; attempt++) {
    const ctx: RetryContext = {
      ...context,
      attempt,
      lastError,
    };

    try {
      return await operation(ctx);
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));

      // Check if we should retry
      if (!isRetryable(lastError, ctx)) {
        throw lastError;
      }

      // Check if we've exhausted attempts
      if (attempt >= fullConfig.maxAttempts) {
        throw new RetryExhaustedError(
          `Operation failed after ${attempt} attempts`,
          attempt,
          lastError
        );
      }

      // Compute delay and sleep
      const delay = computeRetryDelay(attempt, fullConfig);
      await sleep(delay);
    }
  }

  // Should never reach here, but TypeScript needs it
  throw new RetryExhaustedError(
    `Operation failed after ${fullConfig.maxAttempts} attempts`,
    fullConfig.maxAttempts,
    lastError
  );
}

/**
 * Retry policy builder for fluent API.
 */
export class RetryPolicy {
  private config: Partial<RetryConfig> = {};

  /**
   * Set maximum number of attempts.
   */
  maxAttempts(attempts: number): this {
    this.config.maxAttempts = attempts;
    return this;
  }

  /**
   * Set initial backoff delay in milliseconds.
   */
  initialDelay(ms: number): this {
    this.config.initialDelayMs = ms;
    return this;
  }

  /**
   * Set maximum backoff delay in milliseconds.
   */
  maxDelay(ms: number): this {
    this.config.maxDelayMs = ms;
    return this;
  }

  /**
   * Set backoff multiplier.
   */
  backoffMultiplier(multiplier: number): this {
    this.config.backoffMultiplier = multiplier;
    return this;
  }

  /**
   * Set jitter in milliseconds.
   */
  jitter(ms: number): this {
    this.config.jitterMs = ms;
    return this;
  }

  /**
   * Enable or disable retries on NOT_LEADER errors.
   */
  retryOnNotLeader(enabled: boolean): this {
    this.config.retryOnNotLeader = enabled;
    return this;
  }

  /**
   * Build the retry configuration.
   */
  build(): RetryConfig {
    return { ...this.config };
  }

  /**
   * Execute an operation with this retry policy.
   */
  async execute<T>(
    operation: (ctx: RetryContext) => Promise<T>,
    context: Omit<RetryContext, 'attempt'>
  ): Promise<T> {
    return withRetry(operation, this.build(), context);
  }
}

/**
 * Create a new retry policy builder.
 */
export function retryPolicy(): RetryPolicy {
  return new RetryPolicy();
}
